buddy_alloc.h

```diff
@@ -1504,34 +1653,52 @@ static void update_parent_chain(struct buddy_tree *t, struct buddy_tree_pos pos,
 }
 
 static struct buddy_tree_pos buddy_tree_find_free(struct buddy_tree *t, uint8_t target_depth) {
+    struct buddy_tree_pos current_pos;
+    uint8_t target_status;
+    size_t current_depth;
+
+#if defined(__KERNEL__) || defined(MODULE)
+    WARN_ON(target_depth <= t->order);
+#else
     assert(target_depth <= t->order);
-    struct buddy_tree_pos current_pos = buddy_tree_root();
-    uint8_t target_status = target_depth - 1;
-    size_t current_depth = buddy_tree_depth(current_pos);
+    assert(target_depth >= 1);
+#endif
+    current_pos = buddy_tree_root();
+    target_status = (uint8_t) (target_depth - 1);
+    current_depth = buddy_tree_depth(current_pos);
     if (buddy_tree_status(t, current_pos) > target_status) {
         return INVALID_POS; /* No position available down the tree */
     }
     while (current_depth != target_depth) {
+        struct buddy_tree_pos left_pos, right_pos;
+        struct internal_position left_internal, right_internal;
+
+#if !defined(__KERNEL__) && !defined(MODULE)
+        /* make sure no overflow */
+        assert(target_status >= 1);
+#endif
         /* Advance criteria */
-        target_status -= 1;
+        target_status = (uint8_t)(target_status - 1);
```

Comment:
Would target_status underflow?

Reply:
target_status won't underflow - if the tree cannot find a slot to allocate that will be detected by the check outside of the loop.

otherwise the tree is guaranteed to find a slot.